### PR TITLE
Merge identical slots in EIG + MApiary GUI

### LINKED ISSUE #13265: Server crash accessing industrial greenhouse GUI after upgrading its power input
### Your GTNH Discord Username

_No response_

### Your Pack Version

2.3.2

### Your Server

Private Server

### Java Version

Java 19

### Type of Server

Vanilla Forge

### Your Actions

We upgraded an industrial greenhouse to double UMV and accessed its GUI.
Putting it on a single UMV hatch seems to make it functional again.

### Crash Report

[crash-2023-04-20_09.12.55-server.txt](https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/files/11287236/crash-2023-04-20_09.12.55-server.txt)

Stacktrace only:
```java
java.lang.IllegalArgumentException: Dynamic Synced widget id must be greater than 0 and smaller than 65535 (0xFFFF)
	at com.gtnewhorizons.modularui.api.screen.ModularWindow.addDynamicSyncedWidget(ModularWindow.java:416)
	at com.gtnewhorizons.modularui.common.widget.ChangeableWidget.lambda$initQueuedChild$1(ChangeableWidget.java:103)
	at com.gtnewhorizons.modularui.api.widget.IWidgetParent.lambda$forEachByLayer$1(IWidgetParent.java:112)
	at com.gtnewhorizons.modularui.api.widget.IWidgetParent.forEachByLayer(IWidgetParent.java:84)
	at com.gtnewhorizons.modularui.api.widget.IWidgetParent.forEachByLayer(IWidgetParent.java:71)
	at com.gtnewhorizons.modularui.api.widget.IWidgetParent.forEachByLayer(IWidgetParent.java:67)
	at com.gtnewhorizons.modularui.api.widget.IWidgetParent.forEachByLayer(IWidgetParent.java:61)
	at com.gtnewhorizons.modularui.api.widget.IWidgetParent.forEachByLayer(IWidgetParent.java:111)
	at com.gtnewhorizons.modularui.common.widget.ChangeableWidget.initQueuedChild(ChangeableWidget.java:107)
	at com.gtnewhorizons.modularui.common.widget.ChangeableWidget.notifyChangeNoSync(ChangeableWidget.java:76)
	at kubatech.tileentity.gregtech.multiblock.GT_MetaTileEntity_ExtremeIndustrialGreenhouse.lambda$addUIWidgets$11(GT_MetaTileEntity_ExtremeIndustrialGreenhouse.java:625)
	at com.gtnewhorizons.modularui.common.widget.FakeSyncWidget.detectAndSendChanges(FakeSyncWidget.java:64)
	at com.gtnewhorizons.modularui.api.screen.ModularWindow.serverUpdate(ModularWindow.java:226)
	at com.gtnewhorizons.modularui.common.internal.wrapper.ModularUIContainer.lambda$detectAndSendChanges$2(ModularUIContainer.java:118)
	at com.gtnewhorizons.modularui.api.screen.ModularUIContext.forEachWindowTopToBottom(ModularUIContext.java:98)
	at com.gtnewhorizons.modularui.common.internal.wrapper.ModularUIContainer.func_75142_b(ModularUIContainer.java:115)
	at net.minecraft.entity.player.EntityPlayerMP.localOnUpdate(EntityPlayerMP.java:212)
	at api.player.server.ServerPlayerAPI.onUpdate(ServerPlayerAPI.java:6101)
	at net.minecraft.entity.player.EntityPlayerMP.func_70071_h_(EntityPlayerMP.java)
	at net.minecraft.world.World.func_72866_a(World.java:2070)
	at net.minecraft.world.WorldServer.func_72866_a(WorldServer.java:648)
	at net.minecraft.world.World.func_72870_g(World.java:2034)
	at net.minecraft.world.World.func_72939_s(World.java:1887)
	at net.minecraft.world.WorldServer.func_72939_s(WorldServer.java:1146)
	at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:636)
	at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:334)
	at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:547)
	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:396)
	at net.minecraft.server.MinecraftServer$2.run(MinecraftServer.java:685)
```

### Final Checklist

- [X] I have searched this issue tracker and there is nothing similar already. Posting on a closed issue saying the crash still exists will prompt us to investigate and reopen it once we confirm your report.
- [X] I can reproduce this crash consistently by following the exact steps I described above.
- [ ] I have asked other people and they confirm they also crash by following the exact steps I described above.

### PR DESCRIPTION
**Pull request by kuba6000**
Fixes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/13265

### CODE REVIEW

File: src/main/java/kubatech/api/helpers/GTHelper.java [hunk 1]
```diff
@@ -52,4 +60,42 @@ public static int getVoltageTier(long voltage) {
...
+        public void write(PacketBuffer buffer) throws IOException {
+            buffer.writeVarIntToBuffer(count);
+            buffer.writeItemStackToBuffer(stack);
+        }
+
+        public static StackableGUISlot read(PacketBuffer buffer) throws IOException {
+            StackableGUISlot slot = new StackableGUISlot();
+            slot.count = buffer.readVarIntFromBuffer();
+            slot.stack = buffer.readItemStackFromBuffer();
+            return slot;
```
Comment (miozune) [inline src/main/java/kubatech/api/helpers/GTHelper.java]:
```suggestion
            return new StackableGUISlot(
                buffer.readVarIntFromBuffer(),
                buffer.readItemStackFromBuffer(),
                new ArrayList<>());
```
Then `final` can be applied to fields.

File: src/main/java/kubatech/api/helpers/GTHelper.java [hunk 2]
```diff
@@ -52,4 +60,42 @@ public static int getVoltageTier(long voltage) {
...
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) return true;
+            if (!(obj instanceof StackableGUISlot)) return false;
+            return count == ((StackableGUISlot) obj).count && ItemID.createNoCopy(stack, false)
```
Comment (miozune) [inline src/main/java/kubatech/api/helpers/GTHelper.java]: `StackableGUISlot other = (StackableGUISlot) obj` or smth would improve readability

File: src/main/java/kubatech/tileentity/gregtech/multiblock/GT_MetaTileEntity_ExtremeIndustrialGreenhouse.java [hunk 3]
```diff
@@ -625,15 +626,17 @@ public ItemStack transferStackInSlot(EntityPlayer aPlayer, int aSlotIndex) {
                 return super.transferStackInSlot(aPlayer, aSlotIndex);
             }
             if (mte.addCrop(aStack)) {
-                s.putStack(null);
+                if (aStack.stackSize == 0) s.putStack(null);
+                else s.putStack(aStack);
```
Comment (miozune) [inline src/main/java/kubatech/tileentity/gregtech/multiblock/GT_MetaTileEntity_ExtremeIndustrialGreenhouse.java:630]: Is this else clause needed?

File: src/main/java/kubatech/api/helpers/GTHelper.java [hunk 4]
```diff
@@ -52,4 +60,42 @@ public static int getVoltageTier(long voltage) {
     public static int getVoltageTier(GT_MetaTileEntity_MultiBlockBase mte) {
         return (int) getVoltageTierD(mte);
     }
+
+    public static class StackableGUISlot {
```
Comment (miozune) [inline src/main/java/kubatech/api/helpers/GTHelper.java]: I think `StackableItemSlot` would be better, as it's used on server as well.

**Review by miozune — CHANGES_REQUESTED** [review]
- Unlike EEC, MApiary is not saying "empty" for empty slots
- Large number is not displayed well


Reading code, I still think this is a feature-like PR, even if it includes fix. Considering the amount of people affected by the bug (we didn't have ticket for it for a long time), v.s. the amount of people that can be potentially affected by bugs this PR might accidentally introduce (IV+), I don't think it's worth rush merging before stable. Also it's not game-breaking bug, they can downgrade power to workaround crash.

File: src/main/java/kubatech/tileentity/gregtech/multiblock/GT_MetaTileEntity_ExtremeIndustrialGreenhouse.java (hunk 3 above)
Comment (kuba6000) [reply src/main/java/kubatech/tileentity/gregtech/multiblock/GT_MetaTileEntity_ExtremeIndustrialGreenhouse.java:630]: not sure

Comment (miozune) [reply src/main/java/kubatech/tileentity/gregtech/multiblock/GT_MetaTileEntity_ExtremeIndustrialGreenhouse.java:630]: On second thought, this makes sure change to stacksize will be announced, so it's ok

**Review by miozune — APPROVED** [review]

**Pull request merged**